web-chat/src/ChatList.tsx

```diff
@@ -9,23 +14,47 @@ interface Props {
 export default function ChatList(props: Props) {
   const messages = props.messages;
 
-  const listItems = messages.map((currentMessage) => (
-    <Message
-      key={currentMessage.timestamp.toString()}
-      authorName={currentMessage.nick}
-      date={formatDisplayDate(currentMessage)}
-    >
-      <MessageText>{currentMessage.message}</MessageText>
-    </Message>
-  ));
+  const messagesGroupedBySender = groupMessagesBySender(props.messages).map(
+    (currentMessageGroup) => (
+      <MessageGroup onlyFirstWithMeta>
+        {currentMessageGroup.map((currentMessage) => (
+          <Message
+            key={currentMessage.timestamp.toString()}
+            authorName={currentMessage.nick}
+            date={formatDisplayDate(currentMessage)}
+          >
+            <MessageText>{currentMessage.message}</MessageText>
+          </Message>
+        ))}
+      </MessageGroup>
+    )
+  );
 
   return (
-    <MessageGroup>
-      {listItems}
+    <MessageList active containScrollInSubtree>
+      {messagesGroupedBySender}
       <AlwaysScrollToBottom messages={messages} />
-    </MessageGroup>
+    </MessageList>
   );
 
+  function groupMessagesBySender(messageArray: ChatMessage[]) {
+    var currentSender = -1;
+    var lastNick = '';
+    var messagesBySender: ChatMessage[][] = [];
+    var currentSenderMessage = 0;
+
+    for (let currentMessage of messageArray) {
+      if (lastNick !== currentMessage.nick) {
+        currentSender++;
+        messagesBySender[currentSender] = [];
+        currentSenderMessage = 0;
+        lastNick = currentMessage.nick;
+      }
+      messagesBySender[currentSender][currentSenderMessage++] = currentMessage;
+    }
+    return messagesBySender;
```

Comment:
I don't believe this function needs to be defined within `function ChatList`. Can you please extract it for clarity sake?

Reply:
Sure, honestly I didn't really know where it should be scoped, I put it in more for the code to be self-documenting than reused for the moment. I don't think it will be used elsewhere, you want it just at the root level of that file or extracted elsewhere?

Reply:
formatDisplayDate() is also scoped withing that function, arguably it is more widely reusable.

You could also refactor them where you like. JavaScript is very permissive with what it lets you do and I see lots of different philosophies as to how it should best be structured.

Reply:
I took them both out of the ChatList function and put them up at the root of the file, the date formatter maybe should be refactored to an external utility file, the other may also be reused.